File: mobile-app/lib/models/staff_adjustment.dart
/// One reward or one penalty, written down on the day it was earned.
///
/// The payslip has always had a bonus box and a deduction box, both typed
/// at the moment of payment — which is the end of a long month, when
/// nobody remembers who came in late on the 12th. These are the same two
/// figures, recorded when they happen and with the reason attached, so the
/// month's total is arrived at rather than recalled.
class StaffAdjustment {
  const StaffAdjustment({
    required this.id,
    required this.kind,
    required this.kindLabel,
    required this.basis,
    required this.basisLabel,
    required this.value,
    required this.valueFormatted,
    required this.reason,
    required this.occurredOn,
    required this.isSettled,
    required this.isNoteOnly,
    this.userName,
    this.days,
  });

  final int id;

  /// 'reward' or 'penalty'.
  final String kind;
  final String kindLabel;

  /// 'amount', 'days' or 'note'.
  final String basis;

  /// «نیم روز», «بدون کسر», or empty when the figure speaks for itself.
  final String basisLabel;

  /// In the shop's display unit. Zero for a note-only entry, which is a
  /// different thing from an amount that happens to be zero — ask
  /// [isNoteOnly] rather than comparing this to zero.
  final double value;
  final String valueFormatted;

  final String reason;
  final String occurredOn;
  final bool isSettled;
  final bool isNoteOnly;
  final String? userName;
  final double? days;

  bool get isReward => kind == 'reward';

  factory StaffAdjustment.fromJson(Map<String, dynamic> json) => StaffAdjustment(
        id: (json['id'] as num?)?.toInt() ?? 0,
        kind: '${json['kind'] ?? 'reward'}',
        kindLabel: '${json['kind_label'] ?? ''}',
        basis: '${json['basis'] ?? 'amount'}',
        basisLabel: '${json['basis_label'] ?? ''}',
        value: _double(json['value']),
        valueFormatted: '${json['value_formatted'] ?? ''}',
        reason: '${json['reason'] ?? ''}',
        occurredOn: '${json['occurred_on_jalali'] ?? ''}',
        isSettled: json['is_settled'] == true,
        isNoteOnly: json['is_note_only'] == true,
        userName: (json['user'] as Map?)?['name'] as String?,
        days: json['days'] == null ? null : _double(json['days']),
      );

  static double _double(dynamic value) =>
      value is num ? value.toDouble() : double.tryParse('$value') ?? 0;
}

/// What one person's month came to, and what it is made of.
///
/// Two totals kept apart rather than netted: they land in two different
/// boxes on the payslip, and someone who earned a reward and took a
/// penalty in the same month is owed the sight of both.
class AdjustmentPeriod {
  const AdjustmentPeriod({
    required this.periodLabel,
    required this.rewardTotal,
    required this.rewardTotalFormatted,
    required this.penaltyTotal,
    required this.penaltyTotalFormatted,
    required this.items,
  });

  final String periodLabel;
  final double rewardTotal;
  final String rewardTotalFormatted;
  final double penaltyTotal;
  final String penaltyTotalFormatted;
  final List<StaffAdjustment> items;

  bool get isEmpty => items.isEmpty;

  factory AdjustmentPeriod.fromJson(Map<String, dynamic> json) => AdjustmentPeriod(
        periodLabel: '${json['period_label'] ?? ''}',
        rewardTotal: StaffAdjustment._double(json['reward_total']),
        rewardTotalFormatted: '${json['reward_total_formatted'] ?? ''}',
        penaltyTotal: StaffAdjustment._double(json['penalty_total']),
        penaltyTotalFormatted: '${json['penalty_total_formatted'] ?? ''}',
        items: ((json['items'] as List?) ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(StaffAdjustment.fromJson)
            .toList(),
      );
}

/// A member of staff asking to be paid for the month.
///
/// It carries no amount: the wage is what was agreed, less what has been
/// drawn against it, and inviting a figure would start a negotiation over
/// a number the system already knows. [estimatedNet] is what it would come
/// to if it were paid today — a forecast, and labelled as one.
class SalaryRequest {
  const SalaryRequest({
    required this.id,
    required this.periodLabel,
    required this.status,
    required this.statusLabel,
    required this.requestedOn,
    required this.daysWaiting,
    this.userName,
    this.note,
    this.decisionNote,
    this.estimatedNetFormatted,
  });

  final int id;
  final String periodLabel;
  final String status;
  final String statusLabel;
  final String requestedOn;
  final int daysWaiting;
  final String? userName;
  final String? note;
  final String? decisionNote;
  final String? estimatedNetFormatted;

  bool get isPending => status == 'pending';

  bool get wasRejected => status == 'rejected';

  factory SalaryRequest.fromJson(Map<String, dynamic> json) => SalaryRequest(
        id: (json['id'] as num?)?.toInt() ?? 0,
        periodLabel: '${json['period_label'] ?? ''}',
        status: '${json['status'] ?? 'pending'}',
        statusLabel: '${json['status_label'] ?? ''}',
        requestedOn: '${json['requested_on_jalali'] ?? ''}',
        daysWaiting: (json['days_waiting'] as num?)?.toInt() ?? 0,
        userName: (json['user'] as Map?)?['name'] as String?,
        note: json['note'] as String?,
        decisionNote: json['decision_note'] as String?,
        estimatedNetFormatted: json['estimated_net_formatted'] as String?,
      );
}

File: backend/app/Models/StaffAdjustment.php
<?php

namespace App\Models;

use App\Models\Concerns\BelongsToBakery;
use App\Models\Concerns\RecordsAudit;
use Illuminate\Database\Eloquent\Model;

/**
 * One reward or one penalty, written down on the day it was earned.
 *
 * The payslip's «پاداش» and «کسورات» boxes are typed at payday, which is
 * the end of a long month. Nobody remembers who was late on the 12th. This
 * is the same two figures, recorded when they happen and with the reason
 * attached, so the month's total is arrived at rather than recalled.
 *
 * It is not itself money moving. Nothing leaves an account here — the
 * payslip does that, once, for the net. This only says what the month
 * came to.
 */
class StaffAdjustment extends Model
{
    use BelongsToBakery, RecordsAudit;

    public const REWARD = 'reward';

    public const PENALTY = 'penalty';

    /** A sum, straight out. */
    public const BY_AMOUNT = 'amount';

    /** Days of pay, priced from this person's own monthly wage. */
    public const BY_DAYS = 'days';

    /** On the record, worth nothing. Saying it was the point. */
    public const BY_NOTE = 'note';

    /**
     * Days in a month for pricing a day of pay.
     *
     * Thirty, not the count of days actually worked: a wage here is agreed
     * as a monthly figure and is not docked for the shop being shut. Using
     * working days would make the same half-day cost more in a short month,
     * which is not what anyone agreed to.
     */
    public const DAYS_IN_MONTH = 30;

    protected $fillable = [
        'user_id',
        'recorded_by',
        'kind',
        'basis',
        'amount',
        'days',
        'occurred_on',
        'reason',
        'salary_payment_id',
    ];

    protected function casts(): array
    {
        return [
            'occurred_on' => 'date',
            'amount' => 'decimal:2',
            'days' => 'decimal:2',
        ];
    }

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function recordedBy()
    {
        return $this->belongsTo(User::class, 'recorded_by');
    }

    public function salaryPayment()
    {
        return $this->belongsTo(SalaryPayment::class);
    }

    /**
     * What this is worth in money, in Toman.
     *
     * A days-based entry is priced here rather than at the moment it was
     * recorded, so a wage agreed later still prices it correctly — and a
     * note-based one is worth nothing by design, which is different from
     * an amount that happens to be zero.
     */
    public function getValueAttribute(): float
    {
        return match ($this->basis) {
            self::BY_AMOUNT => round((float) $this->amount, 2),
            self::BY_DAYS => round((float) $this->days * $this->dailyWage(), 2),
            default => 0.0,
        };
    }

    public function dailyWage(): float
    {
        $monthly = (float) ($this->user?->monthly_salary ?? 0);

        return $monthly <= 0 ? 0.0 : round($monthly / self::DAYS_IN_MONTH, 2);
    }

    public function getIsRewardAttribute(): bool
    {
        return $this->kind === self::REWARD;
    }

    /** Nothing to pay or dock — it is here so it is not forgotten. */
    public function getIsNoteOnlyAttribute(): bool
    {
        return $this->basis === self::BY_NOTE;
    }

    public function getKindLabelAttribute(): string
    {
        return $this->is_reward ? 'تشویقی' : 'تنبیهی';
    }

    /** How it reads on a list: «نیم روز» or a sum or nothing at all. */
    public function getBasisLabelAttribute(): string
    {
        return match ($this->basis) {
            self::BY_DAYS => (float) $this->days === 0.5
                ? 'نیم روز'
                : rtrim(rtrim(number_format((float) $this->days, 2), '0'), '.').' روز',
            self::BY_NOTE => 'بدون کسر',
            default => '',
        };
    }

    public function scopeRewards($query)
    {
        return $query->where('kind', self::REWARD);
    }

    public function scopePenalties($query)
    {
        return $query->where('kind', self::PENALTY);
    }

    /** Not yet carried onto any payslip. */
    public function scopeUnsettled($query)
    {
        return $query->whereNull('salary_payment_id');
    }

    /**
     * What one person's month comes to, as a reward total and a penalty
     * total kept apart.
     *
     * Apart, because they land in two different boxes on the payslip and
     * netting them off would hide both. Someone who earned a reward and
     * took a penalty in the same month is owed the sight of both.
     */
    public static function monthFor(int $userId, $from, $until): array
    {
        $rows = static::with('user:id,monthly_salary')
            ->where('user_id', $userId)
            ->unsettled()
            ->whereBetween('occurred_on', [$from, $until])
            ->get();

        return [
            'reward' => round($rows->where('kind', self::REWARD)->sum('value'), 2),
            'penalty' => round($rows->where('kind', self::PENALTY)->sum('value'), 2),
            'count' => $rows->count(),
        ];
    }
}
